jq255.md

```diff
@@ -0,0 +1,1425 @@
+
+  - The signer's public key is the group element `Q = sk*G`.
+
+#### Input Message
+
+The input message can be *raw* (an arbitrary sequence of bytes) or
+*pre-hashed* (a hash value). In all operations below, we use the
+*prepared message* `M` defined as follows:
+
+  - If the input is raw, then `M` consists of a single byte of value
+    `0x52`, followed by the input bytes.
+
+  - If the input is a hash value `hv`, then `M` is obtained as the
+    concatenation of, in that order:
+
+      - A single byte of value `0x48`.
+
+      - The symbolic name of the hash function that was used to produce
+        `hv`, encoded in ASCII.
+
+      - A single byte of value `0x00` (which terminates the encoded hash
+        function name).
+
+      - The have value `hv`.
+
+The following symbolic hash function names are defined:
+
+| Hash function | symbolic name |
+| :------------ | :------------ |
+| SHA-256       | sha256        |
+| SHA-384       | sha384        |
+| SHA-512       | sha512        |
+| SHA-512/226   | sha512256     |
+| SHA3-256      | sha3256       |
+| SHA3-384      | sha3384       |
+| SHA3-512      | sha3512       |
+| BLAKE2s       | blake2s       |
+| BLAKE2b       | blake2b       |
+| BLAKE3        | blake3        |
+
+In the table above, "BLAKE2s" and "BLAKE2b" are to be understood as
+"BLAKE2s with a 32-byte (256-bit) output" and "BLAKE2b with a 64-byte
+(512-bit) output", respectively. In general, the symbolic name of a hash
+function is obtained by converting the hash function human-readable name
+to lowercase and removing all characters which are neither letters or
+digits.
+
+```python
+def prepare_message_raw(msg):
+    return b'\x52' + msg
+
+def prepare_message_prehashed(hv, hashname):
+    return b'\x48' + hashname.encode() + b'\x00' + hv
+```
+
+The **recommended usage** is to apply jq255e and jq255s signatures on
+input messages pre-hashed with BLAKE2s with a 32-byte (256-bit) output.
+The signature scheme shall be designated, formally, as follows:
+
+  - **jq255e** and **jq255s**: signatures are computed over the
+    input message pre-hashed with BLAKE2s-256 (i.e. BLAKE2s with a
+    32-byte output).
+
+  - **jq255e-raw** and **jq255s-raw**: signatures are computed over
+    the raw input message, with no pre-hashing.
+
+  - **jq255e-hashname** and **jq255s-hashname**: signatures are computed
+    over the input message pre-hashed with the hash function whose
+    *symbolic name* is "hashname". Thus, "jq255e-sha256" means
+    "signature with the jq255e group, computed over an input message
+    pre-hashed with SHA-256".
+
+Thus, "jq255e", as a signature scheme name, is equivalent to
+"jq255e-blake2s".
+
+The following remarks apply:
+
+  - Regardless of the hash function used to pre-hash the message (if any),
+    the *internal* hash function used to compute the challenge is always
+    BLAKE2s. Within the scope of this specification, the use of BLAKE2s
+    internally is not configurable.
```

Comment:
oof that might get confusing

Reply:
I am trying to decouple the "outside" hash function for pre-hashing, on which I have little control, from the "inside" hash function, which is not visible in the API. The idea is that it should be possible to process the bulk of the data with anything that has good performance (e.g. BLAKE3, if that's your thing, or SHA3 if you are a die-hard NIST fan) without requiring the signature engine itself to support gazillions of possible hash functions. A useful model here is that of a smartcard or a HSM: the bulk hashing is done externally, in the calling machine, and the smartcard or HSM only receives the resulting hash value.